proposal/shared-resources.d.ts.md

```diff
@@ -0,0 +1,247 @@
+
+[shared state]: https://github.com/sass/sass/issues/3296
+
+This also means that the proposal makes no assertions about whether file content
+has changed. It is also up to the user to determine when to start and stop a
+long-running compiler process.
+
+### Example
+
+#### Sync Compiler
+
+```js
+import * as sass from 'sass';
+function setup() {
+  const compiler = sass.initCompiler();
+  const result1 = compiler.compileString('a {b: c}').css;
+  const result2 = compiler.compileString('a {b: c}').css;
+  compiler.dispose();
+
+  // throws error
+  const result3 = sass.compileString('a {b: c}').css;
+}
+```
+
+#### Async Compiler
+
+```js
+import * as sass from 'sass';
+async function setup() {
+  const compiler = await sass.initAsyncCompiler();
+  const result1 = await compiler.compileStringAsync('a {b: c}').css;
+  const result2 = await compiler.compileStringAsync('a {b: c}').css;
+  compiler.dispose();
+
+  // throws error
+  const result3 = sass.compileStringAsync('a {b: c}').css;
+}
+```
+
+## API
+
+### Types
+
+```ts
+import {CompileResult} from '../spec/js-api/compile';
+import {Options, StringOptions} from '../spec/js-api/options';
+```
+
+#### initCompiler()
+
+Returns a synchronous [Compiler].
+
+[Compiler]: #compiler
+
+```ts
+export function initCompiler(): Compiler;
+```
+
+#### initAsyncCompiler()
+
+Resolves with an [Async Compiler].
+
+[Async Compiler]: #async-compiler
+
+```ts
+export function initAsyncCompiler(): Promise<AsyncCompiler>;
+```
+
+#### Compiler
+
+An instance of the synchronous [Compiler]. Re-exports [`compile()`]
+and [`compileString()`].
+
+[`compile()`]: ../spec/js-api/compile.d.ts.md#compile
+[`compilestring()`]: ../spec/js-api/compile.d.ts.md#compilestring
+
+```ts
+export interface Compiler {
+  compile(path: string, options?: Options<'sync'>): CompileResult;
+  compileString(source: string, options?: StringOptions<'sync'>): CompileResult;
+```
+
+##### dispose()
+
+Returns the result of [Disposing a Compiler].
+
+[disposing a compiler]: #disposing-a-compiler
+
+```ts
+  dispose(): Boolean;
+}
+```
+
+#### Async Compiler
+
+An instance of the asynchronous [Compiler interface]. Re-exports
+[`compileAsync()`] and [`compileStringAsync()`].
+
+[`compileasync()`]: ../spec/js-api/compile.d.ts.md#compileasync
+[`compilestringasync()`]: ../spec/js-api/compile.d.ts.md#compilestringasync
+
+```ts
+export interface AsyncCompiler {
+  compileAsync(
+    path: string,
+    options?: Options<'async'>
+  ): Promise<CompileResult>;
+  compileStringAsync(
+    source: string,
+    options?: StringOptions<'async'>
+  ): Promise<CompileResult>;
+```
+
+##### dispose()
+
+* Resolves with the result of [Disposing a Compiler].
+
+```ts
+  dispose(): Promise<Boolean>;
+}
+```
+
+## Semantics
+
+### Compiler
+
+A Compiler must:
+
+* Have a lifetime, which starts upon construction.
+
+* `compile` and `compileString` must have identical semantics to the [Sass
+  interface].
```

Comment:
This should be part of the specifications for these functions.

Reply:
I moved this into the Types section in ff34f2d8e6039df5b0adeeabefdee67edddfbdd5.